csrc/device_lower/validation.cpp

```diff
@@ -206,6 +206,191 @@ void validateCpAsyncBulk(const std::vector<TensorView*>& tvs) {
+}
+
+// Expr-specific validaion
+//
+// TODO: Move individual validations to here, e.g.,
+// validateCpAsyncBulk can be moved here
+class ExprValidator : public OptOutDispatch {
+ public:
+  static void validate(Fusion* fusion) {
+    ExprValidator validator(fusion);
+  }
+
+ private:
+  ExprValidator(Fusion* fusion) {
+    for (auto expr : fusion->exprs()) {
+      dispatch(expr);
+    }
+  }
+
+  // The lowering of ArgsortOp depends on specific scheduling
+  // assumptions. This tight coupling isn't ideal, but it's not a
+  // problem for now.
+  void handle(ArgsortOp* aop) final {
+    auto inp_tv = ir_utils::getTvInput(aop);
+    auto out_tv = ir_utils::getTvOutput(aop);
+
+    // Both input and output must be Local tensors. The local input
+    // tensor is initialized to the max or min value so that
+    // out-of-bounds items do not affect the results
+    NVF_ERROR_EQ(inp_tv->getMemoryType(), MemoryType::Local);
+    NVF_ERROR_EQ(out_tv->getMemoryType(), MemoryType::Local);
+
+    // The input will be initialized for this argsort. To avoid any
+    // potential conflict of initialization, require the input to be
+    // exclusively used by the argsort
+    NVF_ERROR_EQ(inp_tv->uses().size(), 1);
+
+    IterDomain* grouped_id = nullptr;
+    for (const auto& loop_id : out_tv->getLoopDomain()) {
+      if (loop_id->getParallelType() == ParallelType::Group) {
+        NVF_ERROR(grouped_id == nullptr, "Multiple IDs found to be grouped");
+        grouped_id = loop_id;
+      }
+    }
+
+    if (grouped_id != nullptr) {
+      // If it's grouped, it must correspond to the innermost subregion
+      // of the logical argsort ID
+      IterDomain* argsort_logical_id =
+          out_tv->getLogicalDomain().at(aop->dim());
+      NVF_ERROR(
+          isInnermost(argsort_logical_id, grouped_id),
+          "Invalid ID to group: ",
+          grouped_id->toString(),
+          " of ",
+          aop->toString());
+
+      // The output is allocated on registers, so the allocation
+      // domain should be just the same as the loop domain. The
+      // grouped ID must have the unit stride, which means all of the
+      // inner IDs must not contribute to the allocation
+      validateUnitStride(out_tv, out_tv->getLoopDomain(), grouped_id);
+
+      // All of the inputs per thread must be provided to the device
+      // function as a contiguous chunk of memory where each element
+      // has a unit stride within its allocated buffer. More
+      // concretely, the input would be passed to the device function
+      // as follows:
+      //
+      //  // Each thread computes 8 argsort elements
+      //  float T1[8];
+      //  ...
+      //  blockArgsort(..., T1, ...);
+      //
+      // Here, the input is required to have a loop ID that is exactly
+      // mapped with the grouped loop ID of the output, and that
+      // producer loop ID is required to use the Serial parallel
```

Comment:
> producer loop ID is required to use the Serial parallel

I think we only require it to *not* be parallelized. i.e. things like vectorize type should also work?!